docs/source/config.rst
Configuration
=============

Configuration data is specified using ``name = value`` pairs in INI-style text files. A detailed syntax_ description is
available at Python.org. Please note that LetsDNS uses `extended interpolation`_ of values, and that entries in the
[DEFAULT] configuration section are inherited by all other sections.

.. _BIND 9: https://bind9.readthedocs.io/en/latest/
.. _dnspython: https://dnspython.readthedocs.io/en/latest/
.. _dynamic action: dynaction.html
.. _extended interpolation: https://docs.python.org/3/library/configparser.html#interpolation-of-values
.. _Hetzner DNS API: https://dns.hetzner.com/api-docs
.. _nsupdate: https://linux.die.net/man/1/nsupdate
.. _RFC 2136: https://datatracker.ietf.org/doc/html/rfc2136.html
.. _syntax: https://docs.python.org/3/library/configparser.html#supported-ini-file-structure

The option names and values listed below have specific semantics. You
can define custom options, but do take care when choosing names. To
avoid clashes, please prefix your custom options with ``x_`` or ``x-``
(i.e., an upper- or lowercase letter *X* followed by an underscore or
horizontal dash).

- action = *identifier*


  LetsDNS "acts" on all all configuration sections which define an action.
  This is the reason why I recommend **not** defining actions in the [DEFAULT]
  section, unless it is the only section in your configuration set.

  The following action identifiers are available:

  - dane-tlsa

    Create DANE TLSA records using the DNS Update protocol (see `RFC 2136`_).
    If your nameserver is running on the same machine as LetsDNS, or if it is
    accessible over a network connection, using this action is the recommend
    and most convenient way to publish DNS records.

  - dynamic:*module.containing.YourActionClass*

    Import and execute a `dynamic action`_ at runtime. Unless you are faced with
    an untypical or advanced use case, you probably won't need this.

  - hetzner-tlsa

    Create DANE TLSA records using the `Hetzner DNS API`_. This action requires
    an API token for write access.

  - nsupdate-stdout

    Print `nsupdate`_ commands for generating DANE TLSA records to stdout.

- api_token = *string*

  API access token. Required when using the *hetzner-tlsa* action.

- api_url = *URL*

  Optional URL for the Hetzner DNS API (default: https://dns.hetzner.com/api/v1)

- cert_ID_path = */path/to/cert.pem*

  Path to a x509 certificate. Replace *ID* with any unique combination
  of numbers and letters. By varying *ID*, you can define as many
  certificates as you need, limited only by available disk space and
  memory.

  Note that LetsDNS automatically deduplicates TLSA records, which implies
  that the number of records created may be lower than the number of
  cert_ID_path entries in your configuration.

- domain = *example.com*

  Fully qualified name of the domain or subdomain to be operated on.

- hostname = *somehost | .*

  Either a hostname without domain, or a single dot character. The latter
  denotes the domain apex.

- keyfile = */path/to/key.json*

  JSON file containing the TSIG keys required for DNS access. The file must
  contain a {"textual-dns-name": "base64-encoded-secret"} pair. For example,
  if your `BIND 9`_ nameserver permits access via
  ``key "mykey" { algorithm hmac-sha256; secret "c2VjcmV0Cg=="; };``
  the matching JSON content is ``{"mykey": "c2VjcmV0Cg=="}``.

- key_algorithm = *algorithm-name*

  TSIG algorithm used for the secret key. The specified name MUST be supported
  by the `dnspython`_ library. Typical values are `hmac-sha256`, like in the
  previous BIND 9 example, or `hmac-sha512`. If you do not specify this option,
  the dnspython library will pick an internal default. This default may not match
  your keyfile, possibly resulting in a dnspython error like "dns.tsig.PeerBadKey:
  The peer didn't know the key we used". It is therefore recommended to explicitly
  define an algorithm.

- nameserver = *hostname | ip-address*

  Host name or IP address for your nameserver. Name resolution is
  delegated to your OS.

- tcp_ports = *port [port [...]]*

  List of TCP service port numbers in the range 1-65535, separated with spaces
  or commas, for which TLSA records should be generated. Default: 25. Different
  ports can be used for other services, i.e. port 443 for HTTPS.

- ttl = *seconds*

  DNS record time-to-live, in seconds. Default: 1800. Values of more than one
  hour may cause issues due to prolonged caching, depending on your roll-over
  strategy for certificates.


docs/source/envvar.rst
Environment variables
=====================

The following environment variables are supported by LetsDNS:

- DNS_TIMEOUT_SECONDS

  Timeout for DNS access operations, in seconds. Default: 30.

- LOG_LEVEL

  One of the Python `log levels`_. Default: ERROR. Using DEBUG can
  generate a significant amount of output.

.. _log levels: https://docs.python.org/3/howto/logging.html#logging-levels


letsdns/configuration.py
# Copyright © 2022-2023 Ralph Seichter
#
# This file is part of LetsDNS.
#
# LetsDNS is free software: you can redistribute it and/or modify it under the terms of the GNU
# General Public License as published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# LetsDNS is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
# even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with LetsDNS.
# If not, see <https://www.gnu.org/licenses/>.
import logging
import sys
from configparser import ConfigParser
from configparser import ExtendedInterpolation
from logging import basicConfig
from logging import debug
from typing import List

from letsdns.util import getenv
from letsdns.util import is_sensitive
from letsdns.util import split

LOG_LEVEL = 'LOG_LEVEL'


def log_level(default: str = 'ERROR') -> int:
    """Return logging level, using the LOG_LEVEL environment variable if available.

    Args:
        default: The default log level.
    """
    return getattr(logging, getenv(LOG_LEVEL, default, debug_env=False).upper())


def init_logger() -> None:
    """Initialise logger."""
    try:
        level = log_level()
        basicConfig(datefmt='%Y-%m-%d %H:%M:%S', format='%(asctime)s %(levelname)s %(message)s', level=level)
    except AttributeError as e:  # pragma: no cover
        print(e, file=sys.stderr)
        sys.exit(1)


class Config:
    """Provide access to configuration data."""
    parser: ConfigParser
    active_section: str  # The currently active configuration section

    def dump(self, destination=sys.stdout) -> None:
        """Dump configuration state into a file.

        Args:
            destination: File pointer.
        """
        self.parser.write(destination)

    def init(self, filenames=None) -> None:
        """Initialise object by loading configuration files from disk.
        Nonexisting or unreadable files are silently ignored.

        Args:
            filenames: Either a single string or a list of strings.
        """
        self.parser = ConfigParser(interpolation=ExtendedInterpolation())
        self.parser.read_dict({'DEFAULT': {'nameserver': '127.0.0.1'}})
        if filenames:
            self.parser.read(filenames, encoding='utf-8')

    def get(self, name: str, fallback=None) -> str:
        """Return an optional configuration value or the specified fallback value.

        Args:
            name: Option name.
            fallback: Returned if option is undefined.
        """
        return self.parser.get(self.active_section, name, fallback=fallback)

    def get_mandatory(self, name: str) -> str:
        """Return a mandatory configuration value.
        Raise an exception if option value is undefined.

        Args:
            name: Option name.
        """
        v = self.parser.get(self.active_section, name)
        if is_sensitive(name):
            debug(f'config: {name} = *****')
        else:
            debug(f'config: {name} = {v}')
        return v

    def get_domain(self) -> str:
        """Return the mandatory 'domain' configuration value.

        Raise an exception if 'domain' is undefined.
        """
        return self.get_mandatory('domain')

    def get_ttl(self) -> int:
        """Return the 'ttl' configuration value (seconds, integer)."""
        return int(self.get('ttl', fallback='1800'))

    def get_tcp_ports(self) -> List[str]:
        """Return the optional 'tcp_ports' configuration value."""
        return split(self.get('tcp_ports', fallback='25'))

    def options(self) -> List[str]:
        """Return all options in the active section."""
        return self.parser.options(self.active_section)
